Add tests for GroupData

wbamberg opened the pull request:
This adds a test for GroupData.json. There's no spec for this structure, so I've looked at the macros that use it to try to figure out what kinds of assumptions they make about the data.

I've treated four of the properties - `interfaces`, `methods`, `properties`, and `events` - as mandatory arrays even if they are empty. As far as I can tell only `events` is actually assumed to be present (in [eventref.ejs](https://github.com/mdn/kumascript/blob/master/macros/eventref.ejs#L36)). However, since `properties` is already present in all groups, and `methods` and `interfaces` are present in all but one group ([and one of those is apparently an error](https://github.com/mdn/kumascript/blob/master/macros/GroupData.json#L1438)), well, it seemed simpler and safer to require all four in the tests, and to add the missing ones to the data.

I've tried to require that strings contain only a restricted set of characters, based on what they already contain and what seems sensible. Macros that use this data sometimes [insert it into the HTML output](https://github.com/mdn/kumascript/blob/master/macros/APIRef.ejs#L263) without any sanitation, so for example stuff like this:

```
        "Alarm API": {
            "overview":   [ "Alarm<strong> API"],
            "interfaces": [ "MozAlarmsManager" ],
            "methods":    [ "Navigator.mozSetMessageHandler()"],
            "properties": [ "Navigator.mozAlarms" ],
            "events":     []
        },
```

...would break pages.

I've tried to use the new testing framework to some extent anyway. I've used `expect` instead of `assert`, but I've not restructured the code to enable (the equivalent of) assertion messages by making every assertion a separate test, since that seems like that will make this code more complicated and verbose. I guess if we decide to use https://github.com/mattphillips/jest-expect-message then it's easy to add assertion messages to this code.

wbamberg commented:
@a2sheppy , @chrisdavidmills please let me know if the tests here match your expectation for what the structure of GroupData ought to be.

ExE-Boss commented on macros/GroupData.json:
```diff
@@ -1893,6 +1893,7 @@
                             "AuthenticatorAssertionResponse" ],
+            "methods":    [],
             "properties": [],
             "events":     []
```
These empty arrays are actually unnecessary.

wbamberg replied:
Please see my PR description where I talk about these.

a2sheppy replied:
There is some code out there that fails if these aren't included. I learned that the hard way a couple months ago.

ExE-Boss commented:
See&nbsp;also&nbsp;https://github.com/mdn/data/pull/342, which shows that there are even more errors in&nbsp;the&nbsp;data.

wbamberg commented:
> See also mdn/data#342, which shows that there are even more errors in the data.

Thanks! As far as I can tell the only extra error there is [the one for the Geolocation API](https://github.com/mdn/data/pull/342/files#r246139024), right? I just fixed that one.

I also added tests to check there aren't extra properties, to catch things like this, and in the process found some I wasn't testing: `dictionaries`, `callbacks` and `types`.

chrisdavidmills commented:
Looks OK to me from a brief glance, but I'm really snowed under this week.

@a2sheppy , you looked at this stuff in detail fairly recently. Can you give this a proper review please?

a2sheppy commented:
> > See also [mdn/data#342](https://github.com/mdn/data/pull/342), which shows that there are even more errors in the data.
> 
> Thanks! As far as I can tell the only extra error there is [the one for the Geolocation API](https://github.com/mdn/data/pull/342/files#r246139024), right? I just fixed that one.
> 
> I also added tests to check there aren't extra properties, to catch things like this, and in the process found some I wasn't testing: `dictionaries`, `callbacks` and `types`.

Yes, these are new; we will be moving stuff that was previously jammed into `interfaces` into these over time. Sidebar code will need to be updated to know to look for them. Same for macros that build lists of interface members.

a2sheppy commented on tests/macros/groupdata.test.js:
```diff
@@ -0,0 +1,161 @@
+const permittedCharacters = {
+  group:      /^[\w ()-]+$/,
+  overview:   /^[\w ()-]+$/,
+  interface:  /^[\w.]+$/,
```
Is it worth testing for the first character of an interface name being upper-case, per the rules for naming interfaces?

wbamberg replied:
Thanks @a2sheppy , I've made this change.

a2sheppy commented on tests/macros/groupdata.test.js:
```diff
@@ -0,0 +1,161 @@
+  group:      /^[\w ()-]+$/,
+  overview:   /^[\w ()-]+$/,
+  interface:  /^[\w.]+$/,
+  property:   /^[\w.]+$/,
```
Similarly, property and method names always start with a lower-case letter (except for constructors).

wbamberg replied:
There are many cases where this isn't true in the data, e.g. https://github.com/mdn/kumascript/blob/master/macros/GroupData.json#L6.

We could make the validation a bunch more complex here (split on dots, and validate the pieces separately) but I don't think it's worth it. The main point of this test is to make sure that assumptions made by the macros aren't broken by data updates.

a2sheppy replied:
Fair enough. I was thinking we could protect against certain typos, too, but you're right that that's outside the scope of verifying the structure of the data.

a2sheppy commented on tests/macros/groupdata.test.js:
```diff
@@ -0,0 +1,161 @@
+  dictionary: /^\w+$/,
+  callback:   /^\w+$/,
+  type:       /^\w+$/,
+  guideTitle: /^[\w .,]+$/,
```
Does this allow for colons? Doesn't look like it, but I know we have titles that include them.

wbamberg replied:
It does not, but there aren't any guides in here with colons (or the test would fail). I've chosen to be conservative with things like this: rather than permit any characters that it seems we might want to use, I'm permitting the ones we actually do use. If someone adds a guide that contains a colon, they can update the regex then.

a2sheppy replied:
Makes sense. I do think we need to be sure to note somewhere that we keep track of that the lack of permitting characters in this test isn't meant to suggest a rule about what's allowed in titles and such then.

ExE-Boss commented on tests/macros/groupdata.test.js:
```diff
@@ -0,0 +1,161 @@
+    // overview is optional
+    if (group.overview) {
+      // if present it must contain only the permitted characters
+      expect(permittedCharacters.overview.test(group.overview)).toBe(true);
```
```suggestion
      expect(group.overview).toMatch(permittedCharacters.overview);
```

ExE-Boss commented on tests/macros/groupdata.test.js:
```diff
@@ -0,0 +1,161 @@
+function checkStringArray(strings, permitted) {
+  expect(Array.isArray(strings)).toBe(true);
+  for (let string of strings) {
+    expect(permitted.test(string)).toBe(true);
```
```suggestion
    expect(string).toMatch(permitted)
```

ExE-Boss commented on tests/macros/groupdata.test.js:
```diff
@@ -0,0 +1,161 @@
+function checkProperties(obj, permitted, mandatory) {
+  let props = Object.keys(obj);
+  for (let prop of props) {
+    expect(permitted.includes(prop)).toBe(true);
```
```suggestion
    expect(permitted).toContain(prop);
```

ExE-Boss commented on tests/macros/groupdata.test.js:
```diff
@@ -0,0 +1,161 @@
+    expect(permitted.includes(prop)).toBe(true);
+  }
+  for (let prop of mandatory) {
+    expect(props.includes(prop)).toBe(true);
```
```suggestion
    expect(props).toContain(prop);
```

ExE-Boss commented on tests/macros/groupdata.test.js:
```diff
@@ -0,0 +1,161 @@
+  for (let groupName of groupNames) {
+
+    // the group name contains only the permitted characters
+    expect(permittedCharacters.group.test(groupName)).toBe(true);
```
```suggestion
    expect(groupName).toMatch(permittedCharacters.group);
```

ExE-Boss commented on tests/macros/groupdata.test.js:
```diff
@@ -0,0 +1,161 @@
+        checkProperties(guide, permittedGuideProperties, mandatoryGuideProperties);
+
+        // these are both strings that contain only the permitted characters
+        expect(permittedCharacters.guideTitle.test(guide.title)).toBe(true);
```
```suggestion
        expect(guide.title).toMatch(permittedCharacters.guideTitle);
```

ExE-Boss commented on tests/macros/groupdata.test.js:
```diff
@@ -0,0 +1,161 @@
+
+        // these are both strings that contain only the permitted characters
+        expect(permittedCharacters.guideTitle.test(guide.title)).toBe(true);
+        expect(permittedCharacters.guideUrl.test(guide.url)).toBe(true);
```
```suggestion
        expect(guide.url).toMatch(permittedCharacters.guideUrl);
```

a2sheppy commented:
As far as I'm concerned, once the changes suggested by @ExE-Boss are in, this is good to go. Nice work! 👍

wbamberg commented:
> To address @escattone’s concerns regarding error messages, I suggest using the proper Jest Matchers methods:

This doesn't help much with that problem, but it is much more readable code, so thanks!

> Also, I’d suggest that the file gets renamed to GroupData.test.js to match the capitalisation of GroupData.json.

I haven't done this. There are no standards for this AFAIK, and to the small extent that there is a convention, lowercase seems to be winning.

> Also, please convert the function comments to JSDoc comments.

Done!

ExE-Boss commented:
> This doesn't help much with that problem, but it is much more readable code, so thanks!

It does however output the&nbsp;exact string that is&nbsp;causing the&nbsp;test to&nbsp;fail and the&nbsp;`RegExp` that should match, so&nbsp;you&nbsp;can do&nbsp;a&nbsp;<kbd>Ctrl&nbsp;+&nbsp;F</kbd> and&nbsp;fix the&nbsp;issue.
